compiler/optimizer/VPHandlers.cpp

```diff
@@ -1916,6 +1942,16 @@ TR::Node *constrainAload(OMR::ValuePropagation *vp, TR::Node *node)
             }
          }
       }
+
+   sig = symRef ? symRef->getTypeSignature(len) : NULL;
+   if (sig)
+      {
+      constraint = createTypeHintConstraint(vp, node, sig, len);
+      if (constraint)
+         {
+         vp->addGlobalConstraint(node, constraint);
+         }
+      }
```

Comment:
1. > I think that for parameters, this should be handled via _parmValues and picked up in mergeDefConstraints()

Do you mean moving `createTypeHintConstraint` to `ValuePropagation::getParmValues`? something like `_parmValues[parmIndex++] = createTypeHintConstraint(vp, node, sig, len)`? 

2. > IMO we should only hint when a signature would normally give us a guaranteed bound

Could you elaborate it a bit? How to check if a signature gives a guaranteed bound? 

3. > That still leaves the static case, but that could be moved up into the static load path just above, which already gets the signature. It should be possible to restrict creating a hint to the case where we fail to get a classBlock

Do you mean move it here? 
https://github.com/eclipse/omr/blob/6bbfdb51b9b5b16eb6c3fe98ffe073d0a2b4efa0/compiler/optimizer/VPHandlers.cpp#L1911-L1915

4. > If we already have a usable classBlock, then we'll use it to create a VPResolvedClass constraint, 

The existing code creates `VPClassType` only if `classBlock && jlClass && classBlock != jlClass`. Don't we want to create hint in other cases when `classBlock` is non-NULL?

Reply:
> Do you mean moving `createTypeHintConstraint` to `ValuePropagation::getParmValues`? something like `_parmValues[parmIndex++] = createTypeHintConstraint(vp, node, sig, len)`?

Yes, it would be set up in `getParmValues()` with something like that.

----

> Could you elaborate it a bit? How to check if a signature gives a guaranteed bound?

Most parameter signatures are guaranteed to accurately describe the type of the parameter on entry to the method, but it's possible to generate IL that reuses the parameter with a different type. In OpenJ9, that would be due to slot sharing in the Java bytecode, where the source declares a distinct variable, but javac sees that the parameter is not live anymore and so it reuses the same slot.

To determine whether the signature is still applicable at a particular use, we have to look at the definitions that reach the use. That's what `mergeDefConstraints()` does to decide whether to use `_parmValues`.

----

> Do you mean move it here?

More or less there, yes. However, while writing my original comment here, I noticed a bug in this code, and the first step of the fix has just been merged: #6814. Since that change has been merged first, I'll refine my earlier comment. I said:

> It should be possible to restrict creating a hint to the case where we fail to get a `classBlock`, either because `getClassFromSignature()` returns null, or because of an interface type.

Now I should say instead: It should be possible to restrict creating a hint to the case where we fail to get a `classBlock` *or* we get an erased `classBlock` due to an interface type. So I'd recommend putting it here:
https://github.com/eclipse/omr/blob/4e39d376e9ecb4c47646a9d3e8bdebc1a623f039/compiler/optimizer/VPHandlers.cpp#L1823-L1824

The hint creation should fit in nicely there if the condition is adjusted only slightly, e.g.
```
if (classBlock == NULL || vp->isUnreliableSignatureType(classBlock, erased))
   {
   // Try to find a hint here!
   classBlock = erased;
   }
```
----

> The existing code creates `VPClassType` only if `classBlock && jlClass && classBlock != jlClass`. Don't we want to create hint in other cases when `classBlock` is non-NULL?

I think the modification to my previous statement just above should cover all the cases where we really want to create a hint that isn't already implied by a `VPResolvedClass` constraint that will be created anyway. I don't think that we need to worry about the cases where `jlClass` is null or `classBlock == jlClass`.

Reply:
Addressed in bb6ab93